cpp/src/io/utilities/legacy/parsing_utils.cu

```diff
@@ -155,6 +155,7 @@ cudf::size_type findAllFromSet(const char *h_data, size_t h_size, const std::vec
 				chunk_offset + result_offset, key, d_count.data().get(), positions);
 		}
 	}
+	CUDA_TRY(cudaStreamSynchronize(0));
```

Comment:
Synchronizing stream zero is quite expensive (it is equivalent to `cudaDeviceSynchronize()`). And it's not clear why this synchronize is needed here. In general we are aiming to avoid synchronization inside libcudf in favor of following stream semantics.

Reply:
In the function of `findAllFromSet()`, the kernel function `countAndSetPositions` is launched on the `stream zero`, so it needs to synchronize on the `stream zero`.
As you asked,  why this synchronize is needed, there is no synchronize right now, and `findAllFromSet()` still works, isn't it?
`findAllFromSet()` seemed to work correctly, because  there is an local variable `rmm::device_vector<cudf::size_type> d_count(1, 0);` defined at the begin of `findAllFromSet()` which is referenced by kernel `countAndSetPositions`, so when return from `findAllFromSet()`,  `d_count`'s deconstructor will be called which will call `cudaFree`, this `cudaFree` will be waited until the kernel `countAndSetPositions` finish.
So  there is an implicit synchronize in `findAllFromSet()` which make it _works correctly_. 
But it's not work correctly all time, it only works correctly if I use libcudf to parse the whole  csv file at once.
It would work incorrectly if I use libcudf to parse csv file piece by piece, because the `findAllFromSet()`'s return value is incorrect if there is no synchronize.

Reply:
Since  `d_count[0]` does not take a stream, it  will do a synchronous cudaMemcpy  -- which uses the default stream. Therefore it will have to wait until after the kernel completes, since the kernel is also run on stream 0. Since this copy is synchronous, it will complete before the function returns, which means before `d_count`'s destructor is called. 

I think therefore the `cudaStreamSynchronize(0)` is not necessary.

If you have observed a bug, it would be helpful if you can report the bug with a GitHub issue that includes instructions for reproducing the bug?

Reply:
Looks indeed like a bug to me: if chunk_count != 0, the copy of each iteration will stomp all over the d_chunk.data() of the previous iteration. It seems like the cudaStreamSynchronize(0) should actually be within the ci loop.

Reply:
It seems that the problematic statement is "return d_count[0];": this uses the host to index a rmm_device_vector device array - not sure how rmm implements it or if it just assumes UVM.
Generally, I'd prefer an explicit cudaMemcpyDtoH to a local host variable here and return that.

Reply:
```suggestion
	cudf::size_type result;
    CUDA_TRY(cudaMemcpy(&result, d_count.data(), sizeof(result), cudaMemcpyDeviceToHost));
    return result;
```

Reply:
yes, I agree with you, explicit `cudaMemcpyDtoH` is better than `cudaStreamSynchronize(0)`.
but, I think result should be copied from `d_count.data().get()` not from `d_count.data()`

Reply:
Oops, yes. data().get(). It still would be root cause why we see a race condition as rmm_device_vector is just a thrust::device_vector, and device_vector::operator[] is supposed to be doing a cudaMemcpy, according to the documentation.

Reply:
@OlivierNV `rmm::device_vector` is just a `thrust::device_vector` with a custom allocator. I believe `thrust::device_vector` `operator[]` called from the host calls cudaMemcpy.

Reply:
@harrism  yes,you are right, however 'thrust::device_vector' 'operator[]' doesn't calls `cudaMemcpy`, it calls `cudaMemcpyAsync`.

this is how `thrust::device_vector` copy device to host, in file `/usr/local/cuda/include/thrust/system/cuda/detail/util.h`
```cpp
template <class Type>
THRUST_HOST_FUNCTION cudaError_t
trivial_copy_from_device(Type *       dst,
                         Type const * src,
                         size_t       count,
                         cudaStream_t stream)
{
  cudaError status = cudaSuccess;
  if (count == 0) return status;

  status = ::cudaMemcpyAsync(dst,
                             src,
                             sizeof(Type) * count,
                             cudaMemcpyDeviceToHost,
                             stream);
  cudaStreamSynchronize(stream);
  return status;
}
```
so, `thrust` would  call `cudaStreamSynchronize` which is **very expensive**.
so, explicit cudaMemDToH is better.

Reply:
What stream is passed to `trivial_copy_from_device`? Since `operator[]` can't take a stream, I assume it is stream 0. Therefore, I think cudaMemcpyAsync + cudaStreamSynchronize would be the same cost as an explicit `cudaMemcpy` (non-async). 

Regardless, as I wrote above this code is legacy code that will be going away soon. @OlivierNV @mjsamoht does the new cuIO code have this same function?

Reply:
Yeah, it looks like findAllFromSet is still being used in the libcudf++ csv reader, in gather_row_offsets().
However, I'm thinking of completely gutting it since the current implementation is quite inefficient, so all this would go away (it appears to do a findAllFromSet followed by a thrust::sort).

Reply:
@OlivierNV so shouldn't we still fix this then until we find time to redo the CSV reader?

Reply:
It would be nice to add a stream parameter (default to zero) to this function and use it within for all operations.

Reply:
@harrism so should I close this #PR and open another one where add a stream parameter to this function, or you have already done this?

Reply:
> What stream is passed to `trivial_copy_from_device`? Since `operator[]` can't take a stream, I assume it is stream 0. Therefore, I think cudaMemcpyAsync + cudaStreamSynchronize would be the same cost as an explicit `cudaMemcpy` (non-async).
> 
> Regardless, as I wrote above this code is legacy code that will be going away soon. @OlivierNV @mjsamoht does the new cuIO code have this same function?

@harrism , reference to [nvidia document](https://docs.nvidia.com/cuda/cuda-runtime-api/stream-sync-behavior.html#stream-sync-behavior)
**The legacy default stream is an implicit stream which synchronizes with all other streams in the same CUcontext except for non-blocking streams, described below.**
I believe `cudaMemcpy` is better than `cudaMemcpyAsync + cudaStreamSynchronize` on the default stream.
However, it need to look into the driver's source code, to check if I'm right.

Reply:
It's not clear to me what the problem is yet, but it may be related to per-thread default stream behavior: I see a rmm::device_buffer being created without a stream with some functions using streams, other using stream0, so it might be that there is a missing sync somewhere, but I suspect it's outside of this function.
@nezha2017 did you see the race condition with the legacy code (not using streams afaik), or in the libcudf++ path ?

Reply:
@OlivierNV ok, I'm sorry to make this confusing, there is no problem in function `findAllFromSet` at all.

Reply:
@nezha2017 are you saying you never experienced a race condition?

Reply:
@harrism ,OK
let me make it clear.
1. I'm using libcudf++ to parse csv file in my recently project.
2. I noticed that if I use libcudf++ to parse the whole csv file at once, libcudf++ works correctly.
3. However if I need to parse a very large csv file, nearly 15G, I need to parse it peace by peace, libcudf++ will crash randomly. It will crash at compute B but work correct at compute B, It works correctly today, but it crashed yesterday on the same compute. It really makes me crazy.
4. Then I looked into the source code of libcudf++, try to fix the `bug`.
5.  Then I notice the `bug` on function `findAllFromSet`, where I thought rmm::device_array is just an array of UVM, so I thought it was an `bug`, then propose this #PR
6. After talking with you, I realized the `bug` is not in `findAllFromSet`.
7. That's all.

Regardless,
I still believe `cudaMemcpy` is better than `cudaMemcpyAsync + cudaStreamSynchronize`, however, may be this is the bug of `thrust` no the `libcudf++`, and may be I should not talk this ideal in this thread.

Reply:
OK, thanks. I'm not sure why you think that cudaMemcpy is better than cudaMemcpyAsync. The latter allows you to control when synchronization happens, and it only syncs the specified stream, whereas the former syncs the null stream (in other words all streams) AFAIK.

In general if you observe a bug, I recommend opening an issue describing the behavior (with reproduction instructions if possible) for discussion, before suggesting a fix.

Reply:
Ok, I'm not mean `cudaMemcpy` is better than `cudaMemcpyAsync`.
I mean `cudaMemcpy` is better than `cudaMemcpyAsyn + cudaStreamSynchronize` on the `default stream`.
Because the [document](https://docs.nvidia.com/cuda/cuda-runtime-api/stream-sync-behavior.html#stream-sync-behavior) said **synchronize on the stream 0 will synchronizes with all other streams in the same CUcontext except for non-blocking stream**.
since Nvidia don't open source the `Driver Code`, it is hard to prove if I'm right. 
I have asked this question on [Nvidia Forum](https://devtalk.nvidia.com/default/topic/1068147/cuda-programming-and-performance/is-cudamemcpyasync-cudastreamsynchronize-on-default-stream-equal-to-cudamemcpy-non-async-/), and I hope the driver developer would give me an answer.

Reply:
I see. I will try to get an answer directly.